db/design_doc.go

```diff
@@ -86,103 +86,113 @@ func (db *Database) PutDesignDoc(ddocName string, ddoc sgbucket.DesignDoc) (err
-		                      return;
-		                    if ((sync.flags & 1) || sync.deleted)
-		                      return;
-		                    var channels = [];
-		                    var channelMap = sync.channels;
-							if (channelMap) {
-								for (var name in channelMap) {
-									removed = channelMap[name];
-									if (!removed)
-										channels.push(name);
-								}
-							}
-		                    delete doc._sync;
-		                    meta.rev = sync.rev;
-		                    meta.channels = channels;
-
-		                    var _emit = emit;
-		                    (function(){
-			                    var emit = function(key,value) {
-			                    	_emit(key,[channels, value]);
-			                    };
-								(%s) (doc, meta);
-							}());
-							doc._sync = sync;
-						}`
-	viewWrapper_adminViews_xattr = `function(doc,meta) {
-	                    var sync = meta.xattrs._sync;
-	                    if (sync === undefined || meta.id.substring(0,6) == "_sync:")
-	                      return;
-	                    if ((sync.flags & 1) || sync.deleted)
-	                      return;
-	                    meta.rev = sync.rev;
-						(%s) (doc, meta);}`
-	viewWrapper_userViews_xattr = `function(doc,meta) {
-		                    var sync = meta.xattrs._sync;
-		                    if (sync === undefined || meta.id.substring(0,6) == "_sync:")
-		                      return;
-		                    if ((sync.flags & 1) || sync.deleted)
-		                      return;
-		                    var channels = [];
-		                    var channelMap = sync.channels;
-							if (channelMap) {
-								for (var name in channelMap) {
-									removed = channelMap[name];
-									if (!removed)
-										channels.push(name);
-								}
-							}
-		                    meta.rev = sync.rev;
-		                    meta.channels = channels;
-
-		                    var _emit = emit;
-		                    (function(){
-			                    var emit = function(key,value) {
-			                    	_emit(key,[channels, value]);
-			                    };
-								(%s) (doc, meta);
-							}());
-						}`
-)
-
-func getViewWrapper(enableUserViews bool, useXattrs bool) string {
-	if enableUserViews {
-		if useXattrs {
-			return viewWrapper_userViews_xattr
+	syncViewAdminWrapper = `
+	function(doc, meta) {
+	
+		//Skip any internal sync documents
+		if (meta.id.substring(0, 6) == "_sync:") {
+			return;
+		}
+		var sync;
+		var isXattr;
+	
+		//Get sync data from xattrs or from the doc body
+		if (meta.xattrs === undefined || meta.xattrs.%s === undefined) {
+			sync = doc._sync;
+			isXattr = false;
 		} else {
-			return viewWrapper_userViews
+			sync = meta.xattrs.%s;
+			isXattr = true;
 		}
-	} else {
-		if useXattrs {
-			return viewWrapper_adminViews_xattr
+	
+		//Skip if the document has been deleted or has no sync data defined
+		if ((sync.flags & 1) || sync.deleted || sync === undefined)
```

Comment:
This looks like it needs to be reordered with sync===undefined first to avoid an NPE attempting to reference sync.flags and sync.deleted.

Reply:
Including some negative/additional unit tests for these conditions seems like a good idea, if we don't currently have any coverage.